File: 003-LongestSubstringWithoutRepeatingCharacters/Python/main.py
class Solution:
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        overall_max = 0
        max_len = 0
        d = {}
        
        for idx, letter in enumerate(s):
            if letter not in d:
                max_len += 1
                
            if letter in d:
                max_len = max_len + 1 if idx - d[letter] > max_len else idx - d[letter]
            
            d[letter] = idx
            
            overall_max = max(max_len, overall_max)
            
        return overall_max

File: 003-LongestSubstringWithoutRepeatingCharacters/C++/main.cpp
#include <string>

using namespace std;

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        size_t size = s.size();
        int overall_max = 0, max_len = 0;
        
        map<char, int> d;
        
        for(size_t i = 0; i < size; ++i) {
            if (d.count(s[i]) == 0) {
                max_len++;
            } else {
                max_len = i - d[s[i]] > max_len ? max_len + 1 : i - d[s[i]];
            }
            
            d[s[i]] = i;
            
            overall_max = max_len > overall_max ? max_len : overall_max;
        }
        
        return overall_max;
    }
};